base/linalg/diagonal.jl

```diff
@@ -180,8 +180,12 @@ end
 (\)(D::Diagonal, b::AbstractVector) = reshape(scale(1 ./ D.diag, reshape(b, length(b), 1)), length(b))
 (\)(Da::Diagonal, Db::Diagonal) = Diagonal(Db.diag ./ Da.diag)
 
+invsimilar{T}(d::Vector{T}) = similar(d, T)
+invsimilar{T<:Integer}(d::Vector{T}) = similar(d, Float64)
+invsimilar{T<:Integer}(d::Vector{Complex{T}}) = similar(d, Complex{Float64})
+
 function inv{T}(D::Diagonal{T})
-    Di = similar(D.diag)
+    Di = invsimilar(D.diag)
```

Comment:
Great. Two small details. Eventually, a more rigorous analysis is required, but for now, I just try to take the union of all the element types I can think of when deciding on the right *promotion rule* to use. If `B` is the inverse of `A` then it should be true that `B*A=I`. If the elements of `A` have units like those in `CIUnits` then `one` wouldn't have units whereas `zero` would. Therefore, I think it is better to use `zero`. Dividing with zero shouldn't be a problem for determining the type.